Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/TicketService.hs

```diff
@@ -594,3 +674,244 @@ postTicketBookingsUpdateSeats _ TicketBookingUpdateSeatsReq {..} = do
+    preventFutureTicketBookingServiceStatus visitDate cancelDescription specialDayType serviceCategoryId mId = do
+      QTSM.updateBookedAndBlockedSeats 0 0 serviceCategoryId visitDate
+      mbSplOccasion <- QSO.findBySplDayAndEntityIdAndDate Domain.Types.SpecialOccasion.Closed serviceCategoryId.getId (Just visitDate)
+      case mbSplOccasion of
+        Nothing -> mkSpecialOccasionEntry req.businessHourId (Just visitDate) cancelDescription serviceCategoryId specialDayType (Just mId)
+        Just splOccasion -> unless (elem req.businessHourId splOccasion.businessHours) $ QSO.updateBusinessHoursById (splOccasion.businessHours ++ [req.businessHourId]) splOccasion.id
+
+    mkSpecialOccasionEntry :: Kernel.Types.Id.Id Domain.Types.BusinessHour.BusinessHour -> Maybe Data.Time.Calendar.Day -> Maybe Text -> Kernel.Types.Id.Id Domain.Types.ServiceCategory.ServiceCategory -> Domain.Types.SpecialOccasion.SpecialDayType -> Maybe (Kernel.Types.Id.Id Domain.Types.Merchant.Merchant) -> Environment.Flow ()
+    mkSpecialOccasionEntry bHourId date description serviceCategoryId specialDayType mbMerchantId = do
+      splOccasionId <- generateGUID
+      now <- getCurrentTime
+      let splOccasionEntry =
+            Domain.Types.SpecialOccasion.SpecialOccasion
+              { businessHours = [bHourId],
+                date = date,
+                dayOfWeek = Nothing,
+                description = description,
+                entityId = serviceCategoryId.getId,
+                id = splOccasionId,
+                specialDayType = specialDayType,
+                merchantId = mbMerchantId,
+                merchantOperatingCityId = Nothing,
+                createdAt = now,
+                updatedAt = now
+              }
+      QSO.create splOccasionEntry
+
+    cancelAndRefundConfirmedBookingService :: UTCTime -> PersonMap -> TicketBookingServiceMap -> TicketBookingMap -> DTB.TicketBookingServiceCategory -> Environment.Flow (Environment.Flow (), Environment.Flow ())
+    cancelAndRefundConfirmedBookingService now personMap ticketBookingServiceMap ticketBookingMap tbServiceCategory = do
+      let ticketBookingService' = Map.lookup tbServiceCategory.ticketBookingServiceId ticketBookingServiceMap
+          ticketBooking' = ticketBookingService' >>= (\tbs -> Map.lookup tbs.ticketBookingId ticketBookingMap)
+
+      case (ticketBooking', ticketBookingService') of
+        (Just ticketBooking, Just ticketBookingService) ->
+          if isValidTicketBooking ticketBookingService tbServiceCategory now
+            then do
+              let refundAmount = tbServiceCategory.amount.amount - fromMaybe (HighPrecMoney 0) tbServiceCategory.amountToRefund
+              Redis.whenWithLockRedis (ticketBookingCancellationLockKey ticketBooking.id.getId) 10 $ do
```

Comment:
Let's put this lock just when the case starts

Reply:
it shouldn't matter as we are making update in this block only and we are send refundCall, notificationCall

Reply:
Isn't that exactly why this should be at the top? Refund amount should not be concurrently calculated and initiated by two different APIs